language/analyze/analyze.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2016 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// Command analyze performs sentiment, entity, and syntax analysis
+// on a string of text via the Cloud Natural Language API.
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+
+	language "google.golang.org/api/language/v1beta1"
+)
+
+func main() {
+	if len(os.Args) < 2 {
+		usage("Missing command.")
+	}
+
+	ctx := context.Background()
+	hc, err := google.DefaultClient(ctx, language.CloudPlatformScope)
+	if err != nil {
+		log.Fatal(err)
+	}
+	client, err := language.New(hc)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	text := strings.Join(os.Args[2:], " ")
+	if text == "" {
+		usage("Missing text.")
+	}
+
+	switch os.Args[1] {
+	case "entities":
+		printResp(analyzeEntities(client, text))
+	case "sentiment":
+		printResp(analyzeSentiment(client, text))
+	case "syntax":
+		printResp(analyzeSyntax(client, text))
+	default:
+		usage("Unknown command.")
+	}
+}
+
+func usage(msg string) {
+	fmt.Fprintln(os.Stderr, msg)
+	fmt.Fprintln(os.Stderr, "usage: analyze [entities|sentiment|syntax] <text>")
+	os.Exit(2)
+}
+
+func analyzeEntities(s *language.Service, text string) (*language.AnalyzeEntitiesResponse, error) {
+	req := s.Documents.AnalyzeEntities(&language.AnalyzeEntitiesRequest{
+		Document: &language.Document{
+			Content: text,
+			Type:    "PLAIN_TEXT",
+		},
+		EncodingType: "UTF8",
```

Comment:
...almost. That's what I thought initially too, but that'd be too simple to explain ;)
Conceptually, it's an index into the 'character'. But because of variable-width encodings (and how indexing works in the languages), things get hairy.
So, for UTF8 - yes. Because the lowest unit of 'character' is a byte, it's an index into the byte string.
For UTF16 (eg in javascript), the lowest unit of 'character' is 2 bytes, so it's an index into an array where each slot in the array is 2-bytes. eg if I have the string `var str = "😶abc";`, where `😶` takes up 4 bytes, and `a` takes up 2, the index for `a` would be 2, since `😶` takes up slots 0 and 1.